File: banchus-packet/src/main/java/pe/nanamochi/banchus/protocol/PacketWriter.java
package pe.nanamochi.banchus.protocol;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import pe.nanamochi.banchus.core.ServerPacket;
import pe.nanamochi.banchus.io.BanchoDataWriter;
import pe.nanamochi.banchus.io.IDataWriter;

public class PacketWriter {
  private final IDataWriter writer = new BanchoDataWriter();

  public void writePacket(OutputStream stream, ServerPacket packet) throws IOException {
    writer.writeUint16(stream, packet.getPacketType().getId());
    writer.writeUint8(stream, 0);

    ByteArrayOutputStream body = new ByteArrayOutputStream();
    packet.write(this.writer, body);
    byte[] bodyData = body.toByteArray();

    writer.writeInt32(stream, bodyData.length);
    stream.write(bodyData);
  }
}

File: banchus-packet/src/main/kotlin/pe/nanamochi/banchus/protocol/PacketWriter.kt
package pe.nanamochi.banchus.protocol

import java.io.ByteArrayOutputStream
import java.io.OutputStream
import pe.nanamochi.banchus.core.ServerPacket
import pe.nanamochi.banchus.io.DataWriter
import pe.nanamochi.banchus.io.impl.BanchoDataWriter

class PacketWriter {
    private val writer: DataWriter = BanchoDataWriter()

    fun serialize(packet: ServerPacket): ByteArray =
        ByteArrayOutputStream().use { out ->
            writeSinglePacket(out, packet)
            out.toByteArray()
        }

    fun serializeAll(packets: List<ServerPacket>): ByteArray =
        ByteArrayOutputStream().use { out ->
            packets.forEach { writeSinglePacket(out, it) }
            out.toByteArray()
        }

    private fun writeSinglePacket(out: OutputStream, packet: ServerPacket) {
        val packetData =
            ByteArrayOutputStream().use { body ->
                packet.write(writer, body)
                body.toByteArray()
            }

        writer.writeUint16(out, packet.type.id)
        writer.writeUint8(out, 0u)
        writer.writeInt32(out, packetData.size)
        out.write(packetData)
    }
}
